File: python/template/controllers/controller.py
{%- set used_ops = resources | map(attribute='operations') | sum(start=[]) -%}
{%- set uses_body = ('create' in used_ops) or ('update' in used_ops) or ('replace' in used_ops) -%}
from typing import {% if 'list' in used_ops %}List, {% endif %}Optional
{% if cloud_service == 'Azure Function App' -%}
import azure.functions as func
{%- endif %}
{% if cloud_service == 'GCP Cloud Function' -%}
from flask import Request
{%- endif %}
{% if cloud_service == 'AWS Lambda' and uses_body -%}
import json
from errors import ValidationError
{%- endif %}
from services import (
{%- for resource in resources %}
    {{ resource.name }}Service,
{%- endfor %}
)
from repositories.repository import DEFAULT_LIST_LIMIT
from models import (
{%- for resource in resources %}
{%- if ('get_by_id' in resource.operations) or ('list' in resource.operations) or ('create' in resource.operations) or ('update' in resource.operations) or ('replace' in resource.operations) %}
    {{ resource.name }}Response,
{%- endif %}
{%- if ('create' in resource.operations) or ('update' in resource.operations) or ('replace' in resource.operations) %}
    {{ resource.name }},
{%- endif %}
{%- if 'get_by_id' in resource.operations %}
    {{ resource.name }}IdValidation,
{%- endif %}
{%- endfor %}
)

# Upper bound on a single list page, protecting the datastore from
# pathologically large reads even if a client asks for more.
MAX_LIST_LIMIT = 1000


def _coerce_limit(raw: Optional[str]) -> int:
    try:
        limit = int(raw)
    except (TypeError, ValueError):
        return DEFAULT_LIST_LIMIT
    if limit < 1:
        return DEFAULT_LIST_LIMIT
    return min(limit, MAX_LIST_LIMIT)

{% for resource in resources %}
{%- set uses_body_r = ('create' in resource.operations) or ('update' in resource.operations) or ('replace' in resource.operations) %}
class {{ resource.name }}Controller:
    def __init__(self, service: {{ resource.name }}Service):
        self.service = service
{%- if cloud_service == 'AWS Lambda' and uses_body_r %}

    def _parse_body(self, event: dict) -> dict:
        try:
            body = event.get("body") or "{}"
            if event.get("isBase64Encoded"):
                import base64
                body = base64.b64decode(body).decode("utf-8")
            return json.loads(body)
        except (json.JSONDecodeError, Exception):
            raise ValidationError("Invalid JSON in request body.")
{%- endif %}
{%- if "get_by_id" in resource.operations %}
{%- if cloud_service == 'Azure Function App' %}

    async def get_by_id(self, req: func.HttpRequest) -> {{ resource.name }}Response:
        item_id: str = req.route_params.get('item_id')
        {{ resource.name }}IdValidation(id=item_id)
        return await self.service.get_by_id(item_id)
{%- endif %}
{%- if cloud_service == 'GCP Cloud Function' %}

    async def get_by_id(self, request: Request) -> {{ resource.name }}Response:
        path_parts = request.path.strip('/').split('/')
        item_id: str = path_parts[-1] if len(path_parts) > 0 else None
        {{ resource.name }}IdValidation(id=item_id)
        return await self.service.get_by_id(item_id)
{%- endif %}
{%- if cloud_service == 'AWS Lambda' %}

    async def get_by_id(self, event: dict) -> {{ resource.name }}Response:
        item_id: str = (event.get("pathParameters") or {}).get("item_id")
        {{ resource.name }}IdValidation(id=item_id)
        return await self.service.get_by_id(item_id)
{%- endif %}
{%- endif %}
{%- if "list" in resource.operations %}
{%- if cloud_service == 'Azure Function App' %}

    async def get_list(self, req: func.HttpRequest) -> List[{{ resource.name }}Response]:
        limit = _coerce_limit(req.params.get('limit'))
        return await self.service.get_list(limit)
{%- endif %}
{%- if cloud_service == 'GCP Cloud Function' %}

    async def get_list(self, request: Request) -> List[{{ resource.name }}Response]:
        limit = _coerce_limit(request.args.get('limit'))
        return await self.service.get_list(limit)
{%- endif %}
{%- if cloud_service == 'AWS Lambda' %}

    async def get_list(self, event: dict) -> List[{{ resource.name }}Response]:
        limit = _coerce_limit((event.get("queryStringParameters") or {}).get("limit"))
        return await self.service.get_list(limit)
{%- endif %}
{%- endif %}
{%- if "create" in resource.operations %}
{%- if cloud_service == 'Azure Function App' %}

    async def create(self, req: func.HttpRequest) -> {{ resource.name }}Response:
        item_json: dict = req.get_json()
        item = {{ resource.name }}(**item_json)
        return await self.service.create(item)
{%- endif %}
{%- if cloud_service == 'GCP Cloud Function' %}

    async def create(self, request: Request) -> {{ resource.name }}Response:
        item_json: dict = request.get_json()
        item = {{ resource.name }}(**item_json)
        return await self.service.create(item)
{%- endif %}
{%- if cloud_service == 'AWS Lambda' %}

    async def create(self, event: dict) -> {{ resource.name }}Response:
        item_json: dict = self._parse_body(event)
        item = {{ resource.name }}(**item_json)
        return await self.service.create(item)
{%- endif %}
{%- endif %}
{%- if "update" in resource.operations %}
{%- if cloud_service == 'Azure Function App' %}

    async def update(self, req: func.HttpRequest) -> {{ resource.name }}Response:
        item_id: str = req.route_params.get('item_id')
        item_data: dict = req.get_json()
        item = {{ resource.name }}(**item_data)
        item.id = item_id
        return await self.service.update(item)
{%- endif %}
{%- if cloud_service == 'GCP Cloud Function' %}

    async def update(self, request: Request) -> {{ resource.name }}Response:
        path_parts = request.path.strip('/').split('/')
        item_id: str = path_parts[-1] if len(path_parts) > 0 else None
        item_data: dict = request.get_json()
        item = {{ resource.name }}(**item_data)
        item.id = item_id
        return await self.service.update(item)
{%- endif %}
{%- if cloud_service == 'AWS Lambda' %}

    async def update(self, event: dict) -> {{ resource.name }}Response:
        item_id: str = (event.get("pathParameters") or {}).get("item_id")
        item_data: dict = self._parse_body(event)
        item = {{ resource.name }}(**item_data)
        item.id = item_id
        return await self.service.update(item)
{%- endif %}
{%- endif %}
{%- if "replace" in resource.operations %}
{%- if cloud_service == 'Azure Function App' %}

    async def replace(self, req: func.HttpRequest) -> {{ resource.name }}Response:
        item_id: str = req.route_params.get('item_id')
        item_data: dict = req.get_json()
        item = {{ resource.name }}(**item_data)
        item.id = item_id
        return await self.service.replace(item)
{%- endif %}
{%- if cloud_service == 'GCP Cloud Function' %}

    async def replace(self, request: Request) -> {{ resource.name }}Response:
        path_parts = request.path.strip('/').split('/')
        item_id: str = path_parts[-1] if len(path_parts) > 0 else None
        item_data: dict = request.get_json()
        item = {{ resource.name }}(**item_data)
        item.id = item_id
        return await self.service.replace(item)
{%- endif %}
{%- if cloud_service == 'AWS Lambda' %}

    async def replace(self, event: dict) -> {{ resource.name }}Response:
        item_id: str = (event.get("pathParameters") or {}).get("item_id")
        item_data: dict = self._parse_body(event)
        item = {{ resource.name }}(**item_data)
        item.id = item_id
        return await self.service.replace(item)
{%- endif %}
{%- endif %}
{%- if "delete" in resource.operations %}
{%- if cloud_service == 'Azure Function App' %}

    async def soft_delete(self, req: func.HttpRequest) -> None:
        item_id: str = req.route_params.get('item_id')
        await self.service.soft_delete(item_id)
{%- endif %}
{%- if cloud_service == 'GCP Cloud Function' %}

    async def soft_delete(self, request: Request) -> None:
        path_parts = request.path.strip('/').split('/')
        item_id: str = path_parts[-1] if len(path_parts) > 0 else None
        await self.service.soft_delete(item_id)
{%- endif %}
{%- if cloud_service == 'AWS Lambda' %}

    async def soft_delete(self, event: dict) -> None:
        item_id: str = (event.get("pathParameters") or {}).get("item_id")
        await self.service.soft_delete(item_id)
{%- endif %}
{%- endif %}

{% endfor %}

File: go/template/controllers/controller.go
{%- set used_ops = resources | map(attribute='operations') | sum(start=[]) -%}
{%- set uses_body = ('create' in used_ops) or ('update' in used_ops) or ('replace' in used_ops) -%}
{%- set uses_models = uses_body or ('get_by_id' in used_ops) or ('list' in used_ops) -%}
package controllers

import (
	"context"
	_ "embed"
{%- if uses_body %}
	"encoding/json"
	"io"
{%- endif %}
{%- if uses_models %}

	"{{project_endpoint}}/models"
{%- endif %}
	"{{project_endpoint}}/services"
)

//go:embed schemas/create_request.json
var CreateRequestSchema string

//go:embed schemas/update_request.json
var UpdateRequestSchema string

//go:embed schemas/replace_request.json
var ReplaceRequestSchema string

// Bounds for list pagination, protecting the datastore from unbounded reads.
const (
	DefaultListLimit = 100
	MaxListLimit     = 1000
)

// CoerceLimit clamps a requested page size into the supported range,
// falling back to the default when the value is missing or invalid.
func CoerceLimit(limit int) int {
	if limit < 1 {
		return DefaultListLimit
	}
	if limit > MaxListLimit {
		return MaxListLimit
	}
	return limit
}
{% for resource in resources %}
{%- set r = resource.name %}
{%- set lc = resource.name | to_lower_camel %}
type {{ r }}Controller interface {
{%- if "get_by_id" in resource.operations %}
	Get(ctx context.Context, id string) (*models.{{ r }}Dto, error)
{%- endif %}
{%- if "list" in resource.operations %}
	GetList(ctx context.Context, limit int) ([]models.{{ r }}Dto, error)
{%- endif %}
{%- if "create" in resource.operations %}
	Create(ctx context.Context, body io.Reader) (*models.{{ r }}Dto, error)
{%- endif %}
{%- if "update" in resource.operations %}
	Update(ctx context.Context, id string, body io.Reader) (*models.{{ r }}Dto, error)
{%- endif %}
{%- if "replace" in resource.operations %}
	Replace(ctx context.Context, id string, body io.Reader) (*models.{{ r }}Dto, error)
{%- endif %}
{%- if "delete" in resource.operations %}
	Delete(ctx context.Context, id string) error
{%- endif %}
}

type {{ lc }}Controller struct {
	service         services.{{ r }}Service
	schemaValidator services.SchemaValidator
}

func New{{ r }}Controller(service services.{{ r }}Service, schemaValidator services.SchemaValidator) {{ r }}Controller {
	return &{{ lc }}Controller{service: service, schemaValidator: schemaValidator}
}
{%- if "get_by_id" in resource.operations %}

func (c *{{ lc }}Controller) Get(ctx context.Context, id string) (*models.{{ r }}Dto, error) {
	return c.service.Get(ctx, id)
}
{%- endif %}
{%- if "list" in resource.operations %}

func (c *{{ lc }}Controller) GetList(ctx context.Context, limit int) ([]models.{{ r }}Dto, error) {
	return c.service.GetList(ctx, CoerceLimit(limit))
}
{%- endif %}
{%- if "create" in resource.operations %}

func (c *{{ lc }}Controller) Create(ctx context.Context, body io.Reader) (*models.{{ r }}Dto, error) {
	var req models.Create{{ r }}Request
	if err := json.NewDecoder(body).Decode(&req); err != nil {
		return nil, &models.ValidationError{Message: "Invalid request body."}
	}

	if err := c.schemaValidator.Validate(req, "create_request"); err != nil {
		return nil, err
	}

	return c.service.Create(ctx, req)
}
{%- endif %}
{%- if "update" in resource.operations %}

func (c *{{ lc }}Controller) Update(ctx context.Context, id string, body io.Reader) (*models.{{ r }}Dto, error) {
	var req models.Update{{ r }}Request
	if err := json.NewDecoder(body).Decode(&req); err != nil {
		return nil, &models.ValidationError{Message: "Invalid request body."}
	}

	req.Id = id

	if err := c.schemaValidator.Validate(req, "update_request"); err != nil {
		return nil, err
	}

	return c.service.Update(ctx, req)
}
{%- endif %}
{%- if "replace" in resource.operations %}

func (c *{{ lc }}Controller) Replace(ctx context.Context, id string, body io.Reader) (*models.{{ r }}Dto, error) {
	var req models.Replace{{ r }}Request
	if err := json.NewDecoder(body).Decode(&req); err != nil {
		return nil, &models.ValidationError{Message: "Invalid request body."}
	}

	req.Id = id

	if err := c.schemaValidator.Validate(req, "replace_request"); err != nil {
		return nil, err
	}

	return c.service.Replace(ctx, req)
}
{%- endif %}
{%- if "delete" in resource.operations %}

func (c *{{ lc }}Controller) Delete(ctx context.Context, id string) error {
	return c.service.Delete(ctx, id)
}
{%- endif %}
{% endfor %}